libcudacxx/include/cuda/std/__cuda/semaphore.h

```diff
@@ -23,23 +24,147 @@
 
 _LIBCUDACXX_BEGIN_NAMESPACE_CUDA
 
+void simple_backoff() {
+  int wait_iterations = 1;
+  volatile int ctr = 0;
+  const int max_iterations = 1024;
+
+  for (int i = 0; i < wait_iterations; ++i)
+  {
+    ctr++;
+  }
+  wait_iterations = (wait_iterations * 2 >= max_iterations) ? max_iterations : wait_iterations * 2;
+}
+
+
 template <thread_scope _Sco, ptrdiff_t __least_max_value = INT_MAX>
-class counting_semaphore : public _CUDA_VSTD::__semaphore_base<__least_max_value, _Sco>
+class __fair_semaphore
 {
-  static_assert(__least_max_value <= _CUDA_VSTD::__semaphore_base<__least_max_value, _Sco>::max(), "");
+private:
+  const ptrdiff_t max_;
+  cuda::atomic<ptrdiff_t, _Sco> current;
+  cuda::atomic<ptrdiff_t, _Sco> tickets;
```

Comment:
So I only used two atomic variables here. It might be more intuitive for the programmer/maintainer to have a third one, as I am doing `max_ - tickets.load()` to get the available tickets. But this is an implementation detail.